src/thread_task.rs

```diff
@@ -207,43 +207,36 @@ unsafe fn insert_tasks(
     let mut current_block_34: u64;
     let mut maybe_t = tasks.head;
     while let Some(t) = maybe_t {
-        if tasks[t].type_0 == TaskType::TileEntropy {
-            if tasks[first].type_0 > TaskType::TileEntropy {
-                current_block_34 = 11174649648027449784;
-            } else if tasks[first].sby > tasks[t].sby {
-                current_block_34 = 11174649648027449784;
-            } else {
-                if tasks[first].sby < tasks[t].sby {
-                    insert_tasks_between(c, f, first, last, prev_t, Some(t), cond_signal);
-                    return;
-                }
-                current_block_34 = 15904375183555213903;
+        match (
+            tasks[t].type_0,
+            tasks[t].sby,
+            tasks[first].type_0,
+            tasks[first].sby,
+        ) {
+            // entropy coding precedes other steps
+            (TaskType::TileEntropy, _, f_type, _) if f_type > TaskType::TileEntropy => { /* continue */
             }
-        } else {
-            if tasks[first].type_0 == TaskType::TileEntropy {
+            // both are entropy
+            (_, t_sby, _, f_sby) if f_sby > t_sby => { /* continue */ }
+            (_, t_sby, _, f_sby) if f_sby < t_sby => {
                 insert_tasks_between(c, f, first, last, prev_t, Some(t), cond_signal);
                 return;
             }
-            if tasks[first].sby > tasks[t].sby {
-                current_block_34 = 11174649648027449784;
-            } else {
-                if tasks[first].sby < tasks[t].sby {
-                    insert_tasks_between(c, f, first, last, prev_t, Some(t), cond_signal);
-                    return;
-                }
-                if tasks[first].type_0 as c_uint > tasks[t].type_0 as c_uint {
-                    current_block_34 = 11174649648027449784;
-                } else {
-                    if (tasks[first].type_0 as c_uint) < tasks[t].type_0 as c_uint {
-                        insert_tasks_between(c, f, first, last, prev_t, Some(t), cond_signal);
-                        return;
-                    }
-                    current_block_34 = 15904375183555213903;
-                }
+            (_, _, TaskType::TileEntropy, _) => {
```

Comment:
I'm having a hard time convincing myself that the new control flow here is correct. As an example, in the cases where both task types are `TileEntropy` and both `sby` values are equal, the original C would hit the fallthrough case at the bottom of the loop, whereas I believe it would hit this match arm. Am I misreading this?

Reply:
Yeah, I think you are right. Might have done this too late in the evening :| 

I talked to Stephen and he showed me his break-out-of block approach which is closer to the C and easier to review.